File: codes/matlab/ch20/ell_diagnostic.m
function ell_diagnostic()
%% ell_diagnostic - Chapter 20, Etude 20.9.  Read the coefficients before the error.
% Boyd's Rule-of-Thumb 16: a too-small ell causes |a_n| to flatten early.
% Author: Dr. Denys Dutykh.

    script_dir = fileparts(mfilename('fullpath'));
    addpath(fullfile(script_dir, '..', 'ch07'));
    out_dir = fullfile(script_dir, '..', '..', '..', ...
                       'textbook', 'figures', 'ch20', 'matlab');
    if ~exist(out_dir, 'dir'); mkdir(out_dir); end

    NAVY=[20 45 110]/255; CORAL=[231 76 60]/255; TEAL=[22 160 133]/255;
    ORANGE=[230 126 34]/255; PURPLE=[142 68 173]/255;

    N = 64;  L_full = [0.5 1.0 2.0 4.0 8.0 16.0];
    cs_full = {CORAL, ORANGE, TEAL, PURPLE, NAVY, [139 69 19]/255};
    fig = figure('Position',[100 100 1100 800],'Color','w');

    % ---- (a) Three regimes -----------------------------------------
    subplot(2,2,1); hold on; set(gca, 'YScale','log');
    three_ell = [0.5 2.0 16.0]; three_cs = {CORAL, NAVY, TEAL};
    three_lab = {'small (early flatten)', 'good (clean descent)', ...
                 'large (gentle small-n slope)'};
    for k = 1:3
        a = abs(tbn_coeffs(N, three_ell(k)));
        env = envelope_max(a, 3);
        ns = 0:N;
        plot(ns, env + 1e-18, '-', 'Color', three_cs{k}, 'LineWidth', 1.2, ...
             'DisplayName', sprintf('ell=%g - %s', three_ell(k), three_lab{k}));
        plot(ns(1:4:end), env(1:4:end) + 1e-18, 'o', 'Color', three_cs{k}, ...
             'MarkerFaceColor', 'w', 'MarkerSize', 4, 'HandleVisibility', 'off');
    end
    ylim([1e-17 10]); grid on; box on; xlabel('degree n'); ylabel('envelope of |a_n|');
    title('(a) three regimes of ell (envelope view), N=64');
    legend('Location', 'southwest', 'FontSize', 8);

    % ---- (b) Full sweep, envelope only -----------------------------
    subplot(2,2,2); hold on; set(gca, 'YScale','log');
    for i = 1:length(L_full)
        a = abs(tbn_coeffs(N, L_full(i)));
        env = envelope_max(a, 3);
        plot(0:N, env + 1e-18, '-', 'Color', cs_full{i}, 'LineWidth', 0.9, ...
             'DisplayName', sprintf('ell=%g', L_full(i)));
    end
    ylim([1e-17 10]); grid on; box on; xlabel('degree n'); ylabel('envelope of |a_n|');
    title('(b) full ell sweep at N=64, envelope only');
    legend('Location', 'southwest', 'FontSize', 8, 'NumColumns', 2);

    % ---- (c) Tail sum vs ell at three resolutions ------------------
    subplot(2,2,3);
    Ls = [0.3 0.5 0.7 1 1.5 2 3 4 6 8 12 16 24];
    hold on; set(gca, 'YScale','log', 'XScale','log');
    N_refs = [24 48 96]; ref_cols = {CORAL, TEAL, NAVY};
    for j = 1:3
        errs = zeros(size(Ls));
        for i = 1:length(Ls)
            a = tbn_coeffs(N_refs(j), Ls(i));
            errs(i) = sum(abs(a(floor(N_refs(j)/2)+1:end)));
        end
        plot(Ls, errs, '-o', 'Color', ref_cols{j}, 'LineWidth', 1.0, ...
             'DisplayName', sprintf('N=%d', N_refs(j)));
    end
    grid on; box on; xlabel('ell'); ylabel('sum_{n>N/2} |a_n| (tail size)');
    title('(c) Valley of good ell broadens with N');
    legend('Location', 'best');

    % ---- (d) Valley centre and width vs N --------------------------
    subplot(2,2,4);
    Ns_extra = [24 32 48 64 96 128];
    centres = zeros(size(Ns_extra));
    widths  = zeros(size(Ns_extra));
    for k = 1:length(Ns_extra)
        N_ref = Ns_extra(k);
        errs_N = zeros(size(Ls));
        for i = 1:length(Ls)
            a = tbn_coeffs(N_ref, Ls(i));
            errs_N(i) = sum(abs(a(floor(N_ref/2)+1:end)));
        end
        emin = min(errs_N);
        good = errs_N <= 3.0 * emin;
        centres(k) = exp(mean(log(Ls(good))));
        widths(k)  = max(Ls(good)) / min(Ls(good));
    end
    hold on; set(gca, 'XScale','log', 'YScale','log');
    plot(Ns_extra, centres, '-o', 'Color', NAVY, 'MarkerFaceColor', 'w', ...
         'MarkerSize', 6, 'LineWidth', 1.1, 'DisplayName', 'valley centre');
    plot(Ns_extra, widths, '-s', 'Color', TEAL, 'MarkerFaceColor', 'w', ...
         'MarkerSize', 5, 'LineWidth', 1.0, 'DisplayName', 'valley width factor');
    grid on; box on; xlabel('N'); ylabel('value (log-log)');
    title('(d) broad-valley centre and width vs N');
    legend('Location', 'best');

    set(fig, 'PaperPositionMode', 'auto');
    print(fig, fullfile(out_dir, 'ell_diagnostic.pdf'), '-dpdf');
    print(fig, fullfile(out_dir, 'ell_diagnostic.png'), '-dpng');
    fprintf('[20.9-matlab] figure saved\n');
end

function out = envelope_max(a, win)
    n = numel(a); out = zeros(size(a));
    for i = 1:n
        lo = max(1, i - win); hi = min(n, i + win);
        out(i) = max(a(lo:hi));
    end
end

function a = tbn_coeffs(N, ell)
    [~, x] = cheb_matrix(N);
    y = ell * x ./ sqrt(1 - x.^2);
    fv = zeros(size(x));
    ok = abs(x) < 1 - 1e-12;
    fv(ok) = 1 ./ cosh(y(ok));
    fv = fv(:);
    V = [fv; fv(N:-1:2)];
    A = real(fft(V))/N; A(1) = A(1)/2; A(N+1) = A(N+1)/2;
    a = A(1:N+1);
end

File: codes/julia/ch20/ell_diagnostic.jl
# ell_diagnostic.jl
# Chapter 20: Spectral Methods on Unbounded Intervals
# Computational Etude 20.9: Read the coefficients before the error.
#
# Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)

using CairoMakie
using FFTW
using Printf

include(joinpath(@__DIR__, "..", "ch07", "cheb_matrix.jl"))

function dct1_coeffs(v)
    N = length(v) - 1
    V = vcat(v, reverse(v[2:N]))
    A = real.(fft(V)) ./ N
    A[1] *= 0.5; A[N + 1] *= 0.5
    return A[1:N + 1]
end

function tbn_coeffs(N, ell)
    _, x = cheb_matrix(N)
    fv = zeros(length(x))
    ok = abs.(x) .< 1.0 - 1e-12
    y_ok = ell .* x[ok] ./ sqrt.(1 .- x[ok] .^ 2)
    fv[ok] .= 1.0 ./ cosh.(y_ok)
    return dct1_coeffs(fv)
end

function envelope_max(a::AbstractVector, win::Int = 3)
    n = length(a)
    out = similar(a)
    for i in 1:n
        lo = max(1, i - win); hi = min(n, i + win)
        out[i] = maximum(a[lo:hi])
    end
    return out
end

function run()
    outdir = joinpath(@__DIR__, "..", "..", "..",
                      "textbook", "figures", "ch20", "julia")
    mkpath(outdir)

    NAVY = colorant"#142D6E"; CORAL = colorant"#E74C3C"; TEAL = colorant"#16A085"
    ORANGE = colorant"#E67E22"; PURPLE = colorant"#8E44AD"
    fig = Figure(size=(1100, 800))

    N = 64
    L_full = [0.5, 1.0, 2.0, 4.0, 8.0, 16.0]
    cols_full = [CORAL, ORANGE, TEAL, PURPLE, NAVY, colorant"#8B4513"]

    # ---- (a) Three regimes ----
    ax1 = Axis(fig[1, 1], xlabel="degree n", ylabel="envelope of |a_n|",
               yscale=log10,
               title="(a) three regimes of ell (envelope view), N=64",
               limits=(nothing, (1e-17, 10.0)))
    three = [(0.5, CORAL, "small (early flatten)"),
             (2.0, NAVY, "good (clean descent)"),
             (16.0, TEAL, "large (gentle small-n slope)")]
    for (ell, color, lbl) in three
        a = abs.(tbn_coeffs(N, ell))
        env = envelope_max(a, 3)
        ns = collect(0:N)
        lines!(ax1, ns, env .+ 1e-18; color = color, linewidth = 1.2,
               label = "ell = $ell — $lbl")
        scatter!(ax1, ns[1:4:end], env[1:4:end] .+ 1e-18;
                 color = :white, strokecolor = color, strokewidth = 1.0,
                 markersize = 5)
    end
    axislegend(ax1; position = :lb, labelsize = 8)

    # ---- (b) Full sweep envelope ----
    ax2 = Axis(fig[1, 2], xlabel="degree n", ylabel="envelope of |a_n|",
               yscale=log10,
               title="(b) full ell sweep at N=64, envelope only",
               limits=(nothing, (1e-17, 10.0)))
    for (i, ell) in enumerate(L_full)
        a = abs.(tbn_coeffs(N, ell))
        env = envelope_max(a, 3)
        lines!(ax2, 0:N, env .+ 1e-18; color = cols_full[i], linewidth = 0.9,
               label = "ell = $ell")
    end
    axislegend(ax2; position = :lb, labelsize = 8, nbanks = 2)

    # ---- (c) Tail sum vs ell at three resolutions ----
    Ls = [0.3, 0.5, 0.7, 1.0, 1.5, 2.0, 3.0, 4.0, 6.0, 8.0, 12.0, 16.0, 24.0]
    ax3 = Axis(fig[2, 1], xlabel="ell", ylabel="sum_{n>N/2} |a_n| (tail size)",
               xscale=log10, yscale=log10,
               title="(c) Valley of good ell broadens with N")
    for (N_ref, col, label) in [(24, CORAL, "N=24"), (48, TEAL, "N=48"), (96, NAVY, "N=96")]
        errs = [sum(abs.(tbn_coeffs(N_ref, ell)[(div(N_ref, 2) + 1):end])) for ell in Ls]
        scatterlines!(ax3, Ls, errs; color=col, label=label)
    end
    axislegend(ax3; position = :rb)

    # ---- (d) Valley centre and width vs N ----
    Ns_extra = [24, 32, 48, 64, 96, 128]
    centres = Float64[]; widths = Float64[]
    for N_ref in Ns_extra
        errs_N = [sum(abs.(tbn_coeffs(N_ref, ell)[(div(N_ref, 2) + 1):end])) for ell in Ls]
        emin = minimum(errs_N)
        good_mask = errs_N .<= 3.0 * emin
        push!(centres, exp(sum(log.(Ls[good_mask])) / count(good_mask)))
        push!(widths,  maximum(Ls[good_mask]) / minimum(Ls[good_mask]))
    end
    ax4 = Axis(fig[2, 2], xlabel="N", ylabel="value (log-log)",
               xscale=log10, yscale=log10,
               title="(d) broad-valley centre and width vs N")
    scatterlines!(ax4, Ns_extra, centres; color = NAVY,
                  markercolor = :white, strokecolor = NAVY, strokewidth = 1.0,
                  markersize = 6, linewidth = 1.1, label = "valley centre")
    scatterlines!(ax4, Ns_extra, widths; color = TEAL,
                  marker = :rect, markercolor = :white, strokecolor = TEAL,
                  strokewidth = 1.0, markersize = 5, linewidth = 1.0,
                  label = "valley width factor")
    axislegend(ax4; position = :rt)

    save(joinpath(outdir, "ell_diagnostic.pdf"), fig)
    save(joinpath(outdir, "ell_diagnostic.png"), fig)
    @printf("[20.9-julia] saved figure\n")
end

if abspath(PROGRAM_FILE) == @__FILE__
    run()
end
